test/commands/track/search_test.rb

```diff
@@ -0,0 +1,47 @@
+require "test_helper"
+
+class Badge::CreateTest < ActiveSupport::TestCase
+  test "only include active tracks" do
+    # Create one active and one inactive track
+    track = create :track, active: true
+    create :track, active: false
+
+    assert_equal [track], Track::Search.()
+  end
+
+  test "wildcard search with criteria" do
+    # Create one track with Ruby in the title
+    track = create :track, title: "A Ruby Track"
+    create :track, title: "A JS Track"
+
+    assert_equal [track], Track::Search.(criteria: "ruby")
+  end
+
+  test "status raises without a user" do
+    assert_raises TrackSearchStatusWithoutUserError do
+      Track::Search.(status: :all)
+    end
+  end
+
+  test "status raises unless its valid" do
+    assert_raises TrackSearchInvalidStatusError do
+      Track::Search.(status: :foobar, user: create(:user))
+    end
+  end
+
+  test "status pivots correctly" do
```

Comment:
Maybe have a separate test for each status? Might be clearer that way.

Reply:
So, I disagree 🙂 

I like the fact we have one setup that we can then test works as expected. This links back to the other comment about deleting the comment. If we split this into three tests, then we have to have the setup for each test with models that don't get used, which makes the test more brittle.

In my thinking, having the superset of states, and then checking they filter correctly, is the safest way to check that everything works as expected.

Reply:
This means that this should be a separate object then. It seems that we're testing things too deep.

Reply:
> This means that this should be a separate object then. It seems that we're testing things too deep.

I don't understand what you mean. 

This is just a test that a method is working correctly. The method can take three options and outputs three values as a result. I don't think we're testing anything too deeply here? Could you expand.

I see two options. One is what I have with one test. The second is to replace it with three tests:

```ruby
  test "status with status all" do
    user = create :user
    joined = create :track
    create :user_track, user: user, track: joined
    unjoined = create :track

    assert_equal [joined, unjoined],
                 Track::Search.(status: :all, user: user)
  end

  test "status with status joined" do
    user = create :user
    joined = create :track
    create :user_track, user: user, track: joined
    create :track

    assert_equal [joined],
                 Track::Search.(status: :joined, user: user)
  end

  test "status with status unjoined" do
    user = create :user
    create :track
    create :user_track, user: user, track: joined
    unjoined = create :track

    assert_equal [unjoined],
                 Track::Search.(status: :unjoined, user: user)
  end
```

I don't see how just duplicating the same setup three times makes this test and more robust or clear. Plus adding a fourth state would mean having to remember to add the extra line in 4 places.

Reply:
I think what I'm saying that this should be a separate object, something like a `StatusFilter`. We can add the relevant tests to that just concerning the status. Having 3 different assertions for 3 different states is telling me that it should be on it's own object. This ensures that the setup is robust across them all. I just felt that it may be too deep as when we add a new status (a good possibility), having more than 3 assertions on 1 test seems to be too much.

I might be looking too far ahead in the future though. I'm completely okay with what we have right now :) Just thinking out loud here.

Reply:
OK, sweet. I think I'm ok with it - but if it gets problematic then definitely agree that we could split it out. Thanks :)